File: services/trading-engine/db/repositories/system_settings.py
"""Read/write system_settings from the trading engine.

The trading engine only needs to READ settings (kill switch, thresholds, etc.).
It never modifies settings — that is the API's responsibility.
"""

import json
from typing import Any

import asyncpg


async def get_setting(
    conn: asyncpg.Connection,
    key: str,
) -> dict[str, Any] | None:
    row = await conn.fetchrow(
        "SELECT key, value, description, updated_at, updated_by "
        "FROM system_settings WHERE key = $1",
        key,
    )
    if row is None:
        return None
    return {
        "key": row["key"],
        "value": json.loads(row["value"]) if isinstance(row["value"], str) else row["value"],
        "description": row["description"],
        "updated_at": row["updated_at"],
        "updated_by": row["updated_by"],
    }


async def get_setting_value(
    conn: asyncpg.Connection,
    key: str,
    default: Any = None,
) -> Any:
    setting = await get_setting(conn, key)
    if setting is None:
        return default
    return setting["value"]


async def is_kill_switch_enabled(conn: asyncpg.Connection) -> bool:
    value = await get_setting_value(conn, "trading_kill_switch_enabled", default=True)
    return value is not False


async def get_all_settings(conn: asyncpg.Connection) -> list[dict[str, Any]]:
    rows = await conn.fetch(
        "SELECT key, value, description, updated_at, updated_by "
        "FROM system_settings ORDER BY key"
    )
    return [
        {
            "key": row["key"],
            "value": json.loads(row["value"]) if isinstance(row["value"], str) else row["value"],
            "description": row["description"],
            "updated_at": row["updated_at"],
            "updated_by": row["updated_by"],
        }
        for row in rows
    ]

File: apps/api/src/db/repositories/system-settings.ts
import { Pool, PoolClient } from 'pg';
import { SystemSetting } from '../types';

function mapRow(row: Record<string, unknown>): SystemSetting {
  return {
    key: row.key as string,
    value: row.value as unknown,
    description: row.description as string | null,
    updatedAt: row.updated_at as Date,
    updatedBy: row.updated_by as string | null,
  };
}

export async function getSetting(
  db: Pool | PoolClient,
  key: string,
): Promise<SystemSetting | null> {
  const { rows } = await db.query(
    'SELECT * FROM system_settings WHERE key = $1',
    [key],
  );
  return rows.length ? mapRow(rows[0]) : null;
}

export async function getSettingValue<T = unknown>(
  db: Pool | PoolClient,
  key: string,
): Promise<T | null> {
  const setting = await getSetting(db, key);
  return setting ? (setting.value as T) : null;
}

export async function getAllSettings(
  db: Pool | PoolClient,
): Promise<SystemSetting[]> {
  const { rows } = await db.query(
    'SELECT * FROM system_settings ORDER BY key',
  );
  return rows.map(mapRow);
}

export async function setSetting(
  db: Pool | PoolClient,
  key: string,
  value: unknown,
  updatedBy?: string | null,
): Promise<SystemSetting> {
  const { rows } = await db.query(
    `INSERT INTO system_settings (key, value, updated_by)
     VALUES ($1, $2::jsonb, $3)
     ON CONFLICT (key) DO UPDATE
       SET value      = EXCLUDED.value,
           updated_at = NOW(),
           updated_by = EXCLUDED.updated_by
     RETURNING *`,
    [key, JSON.stringify(value), updatedBy ?? null],
  );
  return mapRow(rows[0]);
}

export async function isKillSwitchEnabled(
  db: Pool | PoolClient,
): Promise<boolean> {
  const value = await getSettingValue<boolean>(db, 'trading_kill_switch_enabled');
  return value !== false;
}
